sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveGroupByAll.scala

```diff
@@ -0,0 +1,110 @@
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.{AGGREGATE, AGGREGATE_EXPRESSION, UNRESOLVED_ATTRIBUTE}
+
+/**
+ * Resolve "group by all" in the following SQL pattern:
+ *  `select col1, col2, agg_expr(...) from table group by all`.
+ *
+ * The all is expanded to include all non-aggregate columns in the select clause.
+ */
+object ResolveGroupByAll extends Rule[LogicalPlan] {
+
+  val ALL = "ALL"
+
+  /**
+   * Returns true iff this is a GROUP BY ALL aggregate. i.e. an Aggregate expression that has
+   * a single unresolved all grouping expression.
+   */
+  private def matchToken(a: Aggregate): Boolean = {
+    if (a.groupingExpressions.size != 1) {
+      return false
+    }
+    a.groupingExpressions.head match {
+      case a: UnresolvedAttribute => a.name.toUpperCase() == ALL
+      case _ => false
+    }
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUpWithPruning(
+    _.containsAllPatterns(UNRESOLVED_ATTRIBUTE, AGGREGATE), ruleId) {
+    // Match a group by with a single unresolved star
+    case a: Aggregate if matchToken(a) =>
+      // Only makes sense to do the rewrite once all the aggregate expressions have been resolved.
+      // Otherwise, we might incorrectly pull an actual aggregate expression over to the grouping
+      // expression list (because we don't know they would be aggregate expressions until resolved).
+      if (a.aggregateExpressions.forall(_.resolved)) {
+        val groupingExprs =
+          a.aggregateExpressions.filter(!_.containsPattern(AGGREGATE_EXPRESSION))
+
+        // If the grouping exprs are empty, this could either be (1) a valid global aggregate, or
+        // (2) we simply fail to infer the grouping columns. As an example, in "i + sum(j)", we will
+        // not automatically infer the grouping column to be "i".
+        if (groupingExprs.isEmpty && a.aggregateExpressions.exists(containsAttribute)) {
+          // Case (2): don't replace the ALL. We will eventually tell the user in checkAnalysis
+          // that we cannot resolve the all in group by.
+          a
+        } else {
+          // Case (1): this is a valid global aggregate.
+          a.copy(groupingExpressions = groupingExprs)
+        }
+      } else {
+        a
+      }
+  }
+
+  /**
+   * Returns true if the expression includes an Attribute outside the aggregate expression part.
+   * For example:
+   *  "i" -> true
+   *  "i + 2" -> true
+   *  "i + sum(j)" -> true
+   *  "sum(j)" -> false
+   *  "sum(j) / 2" -> false
+   */
+  private def containsAttribute(expr: Expression): Boolean = expr match {
+    case _ if AggregateExpression.isAggregate(expr) =>
+      // Don't recurse into AggregateExpressions
+      false
+    case _: Attribute =>
+      true
+      // TODO: do we need to worry about ScalarSubquery here?
```

Comment:
cc @cloud-fan here's a question for you too

Reply:
Good question. I think the idea is, we should only allow non-correlated subquery expression here. The check can be
```
case sub: SubqueryExpression => sub.outerAttrs.nonEmpty
```